pkg/rtc/participant.go

```diff
@@ -963,133 +868,24 @@ func (p *ParticipantImpl) VerifySubscribeParticipantInfo(pID livekit.Participant
-	p.lock.Unlock()
-
+// onTrackSubscribed handles post-processing after a track is subscribed
+func (p *ParticipantImpl) onTrackSubscribed(subTrack types.SubscribedTrack) {
 	if p.params.ClientInfo.FireTrackByRTPPacket() {
 		subTrack.DownTrack().SetActivePaddingOnMuteUpTrack()
 	}
 
-	subTrack.OnBind(func() {
+	subTrack.AddOnBind(func() {
 		if p.TransportManager.HasSubscriberEverConnected() {
 			subTrack.DownTrack().SetConnected()
 		}
 		p.TransportManager.AddSubscribedTrack(subTrack)
+		//p.supervisor.SetSubscribedTrack(subTrack.ID(), subTrack, sourceTrack)
 	})
-
-	if settings != nil {
-		subTrack.UpdateSubscriberSettings(settings)
-	}
-
-	publisherID := subTrack.PublisherID()
-	p.lock.Lock()
-	_, isAlreadySubscribed := p.subscribedTo[publisherID]
-	p.subscribedTo[publisherID] = struct{}{}
-	p.lock.Unlock()
-	if !isAlreadySubscribed && onSubscribedTo != nil {
-		onSubscribedTo(p, publisherID)
-	}
 }
 
-// RemoveSubscribedTrack removes a track to the participant's subscribed list
-func (p *ParticipantImpl) RemoveSubscribedTrack(subTrack types.SubscribedTrack, sourceTrack types.MediaTrack) {
-	p.lock.Lock()
-	if v, ok := p.trackPublisherVersion[subTrack.ID()]; ok && v > subTrack.PublisherVersion() {
-		p.supervisor.ClearSubscribedTrack(subTrack.ID(), subTrack, sourceTrack)
-		p.lock.Unlock()
-		p.params.Logger.Debugw("ignoring remove subscribedTrack from older version",
-			"current", v,
-			"requesting", subTrack.PublisherVersion(),
-			"trackID", subTrack.ID(),
-		)
-		return
-	}
-	p.params.Logger.Debugw("removed subscribedTrack",
-		"publisherID", subTrack.PublisherID(),
-		"publisherIdentity", subTrack.PublisherIdentity(),
-		"trackID", subTrack.ID(), "kind", subTrack.DownTrack().Kind())
-	p.trackPublisherVersion[subTrack.ID()] = subTrack.PublisherVersion()
-
-	delete(p.subscribedTracks, subTrack.ID())
-	p.supervisor.ClearSubscribedTrack(subTrack.ID(), subTrack, sourceTrack)
-
-	// remove from subscribed map
-	numRemaining := 0
-	for _, st := range p.subscribedTracks {
-		if st.PublisherID() == subTrack.PublisherID() {
-			numRemaining++
-		}
-	}
-
-	//
-	// NOTE
-	// subscribedTracksSettings should not be deleted on removal as it is needed if corresponding publisher migrated
-	// LK-TODO: find a way to clean these up
-	//
-
-	if numRemaining == 0 {
-		delete(p.subscribedTo, subTrack.PublisherID())
-	}
-	p.lock.Unlock()
-
+// onTrackUnsubscribed handles post-processing after a track is unsubscribed
+func (p *ParticipantImpl) onTrackUnsubscribed(subTrack types.SubscribedTrack) {
 	p.TransportManager.RemoveSubscribedTrack(subTrack)
-
-	if numRemaining == 0 {
-		//
-		// When a participant leaves OR
-		// this participant unsubscribes from all tracks of another participant,
-		// have to send speaker update indicating that the participant speaker is no long active
-		// so that clients can clean up their speaker state for the leaving/unsubscribed participant
-		//
-		if p.ProtocolVersion().SupportsSpeakerChanged() {
```

Comment:
this has been moved to Room, since it already handles sending updates when we subscribe to someone new.